climada/hazard/tc_surge_bathtub.py

```diff
@@ -0,0 +1,179 @@
+
+CLIMADA is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with CLIMADA. If not, see <https://www.gnu.org/licenses/>.
+
+---
+
+Define TCSurgeBathtub class.
+"""
+
+__all__ = ['TCSurgeBathtub']
+
+import copy
+import logging
+
+import numpy as np
+import rasterio.warp
+import scipy.sparse as sp
+
+from climada.hazard.base import Hazard
+import climada.util.coordinates as coord_util
+
+LOGGER = logging.getLogger(__name__)
+
+HAZ_TYPE = 'TCSurgeBathtub'
+"""Hazard type acronym for this module"""
+
+MAX_DIST_COAST = 50
+"""Maximum inland distance of the centroids in km"""
+
+MAX_ELEVATION = 10
+"""Maximum elevation of the centroids in m"""
+
+DECAY_RATE_M_KM = 0.2
+"""Decay rate of surge in meters per each km"""
+
+
+class TCSurgeBathtub(Hazard):
+    """TC surge heights in m, a bathtub model with wind-surge relationship and inland decay."""
+
+    def __init__(self):
+        Hazard.__init__(self, HAZ_TYPE)
+
+
+    @staticmethod
+    def from_tc_winds(wind_haz, topo_path, inland_decay=True, add_sea_level_rise=0.0):
+        """Compute tropical cyclone surge from input winds.
+
+        Parameters
+        ----------
+        wind_haz : TropCyclone
+            Tropical cyclone wind hazard object.
+        topo_path : str
+            Path to a raster file containing gridded elevation data.
+        inland_decay : bool, optional
+            Surge decays according to distance from coast. Default: True.
+        add_sea_level_rise : float, optional
+            Sea level rise effect in meters to be added to surge height.
+        """
+        centroids = copy.deepcopy(wind_haz.centroids)
+
+        if not centroids.coord.size:
+            centroids.set_meta_to_lat_lon()
+
+        # Select wind-affected centroids which are inside INLAND_MAX_DIST_KM and |lat| < 61
+        if not centroids.dist_coast.size or np.all(centroids.dist_coast >= 0):
+            centroids.set_dist_coast(signed=True, precomputed=True)
+        coastal_msk = (((wind_haz.intensity > 0).sum(axis=0).A1 > 0)
+                       & (centroids.dist_coast < 0)
+                       & (centroids.dist_coast > -MAX_DIST_COAST * 1000)
+                       & (np.abs(centroids.lat) < 61))
+
+        # elevation at coastal centroids, exclude high centroids
+        coastal_centroids_h = coord_util.read_raster_sample(
+            topo_path, centroids.lat[coastal_msk], centroids.lon[coastal_msk])
+        elevation_msk = ((coastal_centroids_h >= 0)
+                         & (coastal_centroids_h <= MAX_ELEVATION + add_sea_level_rise))
+        coastal_msk[coastal_msk] = elevation_msk
+        coastal_centroids_h = coastal_centroids_h[elevation_msk]
+        coastal_idx = coastal_msk.nonzero()[0]
+
```

Comment:
This part I find very hard to follow. 
Could one add maybe a bit more explanation? Like:
`coastal_msk` is a boolean matrix indexing all coastal centroids, 
`elevation_msk` is a boolean matrix indexing all coastal centroids that are additionally within 0 and MAX_elevation + s.l.rise
Explaining the indexing / expansion and reassigning of `coastal_msk` in line 89

Reply:
It is generally considered bad practice to put information in comments that can be easily deferred from the code itself. The fact that `coastal_msk` and `elevation_msk` are boolean arrays is such an information, as well as the constraint values `0` and `MAX_ELEVATION + slr`.

However, I decided to add a comment in my latest commit. Maybe that helps.

Reply:
Well, but if the code is not self-evident, then something is sub-optimal. Either the variable name makes it clear that it is a boolean, or the context or a comment.

Reply:
Yes, "IF the code is not self-evident". That's exactly what I said: Don't put stuff in the comments that is self-evident.

Reply:
yes, ok, I agree. the first instances of coastal_msk and elevation_msk can be figured out relatively easily, so if commenting on that is bad practice, so be it.
But as said, with line 89 `coastal_msk[coastal_msk] = elevation_msk`i really struggle and don't find it very self-evident what the expansion and re-assigning does.

Reply:
Please continue the discussion in the updated code: https://github.com/CLIMADA-project/climada_python/pull/97/files#r531497507